Assertions return boolean

adiSuper94 opened the pull request:
Fixes #2455. Based on #2564.

adiSuper94 commented:
@novemberborn I have edited test cases only for pass(), fail(), is() and not(). 
If this is fine, I'll go ahead and edit the testcases for other assertions and edit the type definitions also.

novemberborn commented on lib/assert.js:
```diff
@@ -286,6 +286,7 @@ class Assertions {
 			},
 			onSuccess: () => {
 				pass();
+				return true;
```
Do we need to return `false` in `onError` or something?

adiSuper94 replied:
yes, we should.

novemberborn reviewed (commented):
@adiSuper94 thanks for redoing the PR.

I've pushed two commits, one for some weird code I spotted, and then the other to return `false` if the messages are illegal.

I think `notThrowsAsync()` should also return a boolean, just because we can. It's only the `throws()` and `throwsAsync()` assertions that return the thrown error *or undefined* (which we haven't typed because it's annoying).

With regards to the tests, I appreciate that the assertion tests are rather complicated. I'm struggling to understand the changes you've made though. Could you elaborate on your approach? For these tests too I'd prefer if we could use `tap`'s assertions to check whether an AVA assertion returned `true` or `false`.

We do have some setup in place now to use AVA to test AVA, but migrating the assertion tests to that should be separate.

sindresorhus commented:
@adiSuper94 Bump :)

adiSuper94 commented:
@sindresorhus  I have been meaning to work on this for the past few weekends. Thanks for the reminder.

adiSuper94 commented:
@novemberborn , sorry for the delay in response.
> 
> I've pushed two commits, one for some weird code I spotted, and then the other to return `false` if the messages are illegal.
> 
Thanks, I guess my editor is acting a little crazy.

> I think `notThrowsAsync()` should also return a boolean, just because we can. It's only the `throws()` and `throwsAsync()` assertions that return the thrown error _or undefined_ (which we haven't typed because it's annoying).
> 
And added returns for `notThrowsAsync()`

> With regards to the tests, I appreciate that the assertion tests are rather complicated. I'm struggling to understand the changes you've made though. Could you elaborate on your approach? For these tests too I'd prefer if we could use `tap`'s assertions to check whether an AVA assertion returned `true` or `false`.
> 
> We do have some setup in place now to use AVA to test AVA, but migrating the assertion tests to that should be separate.
>
The approach i took was rather crude. I have changed it now.  Let me know if it clear now.
I have edited the testcases to not use ava.

novemberborn commented on test-tap/assert.js:
```diff
@@ -120,8 +128,7 @@ function eventuallyFailsWith(t, fn, subset) {
 	lastFailure = null;
-	fn();
-	if (lastFailure) {
+	if (!fn() && lastFailure) {
```
This should check more explicitly for a `false` value:
```suggestion
	if (fn() === false && lastFailure) {
```

novemberborn commented on test-tap/assert.js:
```diff
@@ -146,8 +153,7 @@ function eventuallyFails(t, fn) {
 	lastFailure = null;
-	fn();
-	if (lastPassed) {
+	if (fn() && lastPassed) {
```
And conversely:

```suggestion
	if (fn() === true && lastPassed) {
```

novemberborn reviewed (commented):
Looks good @adiSuper94. Thanks for coming back to this.

I've added two suggestions to make the tests actually look for `false` and `true` values, rather than falsy/truthy. What do you think?

Would you mind updating the type definitions as well?

novemberborn commented:
You might be able to cherry-pick the type definitions from https://github.com/avajs/ava/pull/2601.

adiSuper94 commented:
@novemberborn , Sure.

adiSuper94 commented:
> I've added two suggestions to make the tests actually look for `false` and `true` values, rather than falsy/truthy. What do you think?
> 


I accepted those, suggestions. It actually helped me improve the tests. Thanks!

novemberborn commented on test-tap/assert.js:
```diff
@@ -146,10 +153,10 @@ function eventuallyFails(t, fn) {
+	if (fn() === true && lastPassed) {
 		t.pass();
 	} else {
+		lastFailure = new Error('Assertion failed');
```
You shouldn't need to assign / override this.

novemberborn commented on test-tap/assert.js:
```diff
@@ -1047,6 +1057,7 @@ test('.throws()', gather(t => {
 		assertions.throws(() => {
 			throw new Error('foo');
 		});
+		return true;
```
Why add this return statement? Same below.

novemberborn commented on test-tap/assert.js:
```diff
@@ -120,8 +128,7 @@ function eventuallyFailsWith(t, fn, subset) {
 
 function fails(t, fn) {
 	lastFailure = null;
-	fn();
-	if (lastFailure) {
+	if (fn() === false && lastFailure) {
 		t.pass();
```
Should this be something like:

```js
const retval = fn();
if (lastFailure) {
  t.false(retval)
}
```

And something similar in `passes` below. Combining the variable check with the return value check makes it hard to know what's being tested.

novemberborn commented on index.d.ts:
```diff
@@ -433,7 +509,7 @@ export interface CbExecutionContext<Context = unknown> extends ExecutionContext<
 }
 
-export type ImplementationResult = PromiseLike<void> | Subscribable | void;
+export type ImplementationResult = PromiseLike<void> | Subscribable | boolean | void;
```
This shouldn't change. It's the return type of the function you pass to `test()`.

novemberborn reviewed (commented):
Thanks @adiSuper94. Couple more questions. And I appreciate these assertion tests are rather cumbersome!

novemberborn commented:
Closing in favor of #2696.
